### PR TITLE
Fix behaivour of aws-load-balancer-security-groups annotation

### LINKED ISSUE #79723: Still unable to manage all SGs outside K8S after PR #49445
**What happened**:

* Create SG outside of K8s - a SG rule has a range of ports which are opened via K8s
* Add this into service.beta.kubernetes.io/aws-load-balancer-security-groups
* Ensure load balancer switches from the range back to individual ports, the CIDR will be 0.0.0.0/0 if loadBalancerSourceRanges was removed too

**What you expected to happen**:
* K8S doesn't modify the SG

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.13.7
- Cloud provider or hardware configuration: EKS
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by Elias481**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
The intention of users utilizing the specified by aws-load-balancer-security-groups annotation is typically, that they want to assign an externally managed security to the load balancer and do not want Kubernetes to modify the group.
(Also see summary/fixed issues)

**Which issue(s) this PR fixes**:
Ensures that Security Group specified by aws-load-balancer-security-groups is not modified on setup of Load Balancer anymore.
Fixes #79723, kubernetes/cloud-provider-aws/issues/65
Redesign implementation from PR #62774
Also addresses #49445, #79279

**Special notes for your reviewer**:
If changing return signature of buildELBSecurityGroupList is an issue I would suggest to rename the function and add a wrapper with the old name.

**Release Note**
```release-note
Fix handling of aws-load-balancer-security-groups annotation. Security-Groups assigned with this annotation are no longer modified by kubernetes which is the expected behaviour of most users. Also no unnecessary Security-Groups are created anymore if this annotation is used.
```

### CODE REVIEW

**Comment by redradrat** [comment]
This actually addresses a quite heavy impact issue, as I understand it? It fixes the override of a loadbalancer's sg to allow incoming traffic from 0.0.0.0. 

Is there anything that can be done to help moving this along from the outside?

**Comment by rmt** [comment]
Ping?  We are also affected by this issue.  Who can approve this?

**Comment by wongma7** [comment]
My inclination is that k8s should only modify security groups it creates, and any other behavior is a bug that we need to fix in this PR. But I am new to this issue so please let me know if my understanding is correct.

For a given ELB, there are 4 potential sources of security groups:

1. Created by k8s
2. Global
3. "Replacement"/"override" security groups defined by "service.beta.kubernetes.io/aws-load-balancer-security-groups"
4. "Extra" security groups defined by "service.beta.kubernetes.io/aws-load-balancer-extra-security-groups"

Kubernetes should only modify security groups from source 1, right?

What if there is no global security group and an ELB has "service.beta.kubernetes.io/aws-load-balancer-extra-security-groups"? Then we k8s to modify the security group it created but not the "extra" security groups?

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (wongma7) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: I think we want to return a map[string]bool to make it possible to express that only the kubernetes-created securitygroup should be setup. I don't mind complicating this function's signature as long as it's documented and tested

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws_test.go
Comment (wongma7) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws_test.go]: IMO we need to define the behavior we expect using tests. We need a much more exhaustive set of tests for the various cases and combinations. For example:

- "service.beta.kubernetes.io/aws-load-balancer-security-groups" takes precedence over global
- a security group is created iff neither "service.beta.kubernetes.io/aws-load-balancer-security-groups" nor global exist
- "service.beta.kubernetes.io/aws-load-balancer-extra-security-groups" is always added
- no security groups are ever marked for setup or modifiable EXCEPT the kubernetes-created one

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (Elias481) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: So this is ok I understad as it is exactly doing this (the bool is only true if a SG has been setup because neighter `service.beta.kubernetes.io/aws-load-balancer-security-groups` was specified nor a global `ElbSecurityGroup`.

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws_test.go
Comment (Elias481) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws_test.go]: So You want me to adjust the test cases (or add some) that they meet these requirements?

**Comment by bhagwat070919** [comment]
Any update on this PR ?

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws_test.go
Comment (wongma7) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws_test.go]: If you don't mind. Or I could add them myself in a follow-up, I don't want it to block the PR

**Comment by wongma7** [comment]
needs rebase, pls ping me for review as I want to see this merged!

**Comment by Elias481** [comment]
@wongma7 I did the rebase. For writing the tests I would do, but there are multiple reasons that I cannot do it very soon. So better to merge for now..

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (bhagwat070919) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: We should leave this comment unless its wrong ?

**Review by bhks — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (Elias481) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Hmm code duplication is something else for me. It would not save any line if intermediate variable is used here and the add to list is consolidated into the outer block and for me I did it to have it more readable / understandable what happens in which branch, but yes can change it back if other like the previous way more.

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (Elias481) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: I thought it's duplicate (see next comment) and also unclear which code-change it referrs to.  
Comments regarding changes should go to the commit message only - at least as long the change it referrs to is not visible in the source but only the current version.

Comment (Elias481) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Now it is wrong I assume because the last change on the block just ensures the function returns information whether the main security group it returns is eligible for setup through kubernetes.  
But I could bring it back in if it may help someone...

Comment (bhagwat070919) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: I think we can just modify the comment here as this:

`When we have Security Groups  specified with the ServiceAnnotationLoadBalancerSecurityGroups annotation 
   we don't create a new default Security Groups`

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (bhagwat070919) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: If we think this improves the readability then I am good with that.

Was just thinking the codebase like kubernetes/Kubernetes if we have similar code then it would be too many duplicate coding.
Like in this change itself , the setupSg was set to false and was not updated again in the if block
```
if c.cfg.Global.ElbSecurityGroup != "" {

```
Because its duplicate but would definitely add readability that ok the value is  set to false irrespective of what has been set earlier.

**Review by bhks — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (Elias481) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Ok, so I will insert this comment instead of the both old comments.

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (Elias481) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: I think an append to array line is not something very different to an assignment so no additional duplicated code. But at the end I'd be fine with both ways.

For the setupSg the intention was to show: it's always as default False and should only in explicit case be set to true. Of course we could set it a 2nd time to false in `if c.cfg.Global.ElbSecurityGroup != "" {` block and another time in an else block to the `if len(sgList) == 0 {` but for me this would be too much.

Anyway at the end I'm fine to change the code style, I just want the fix :-D

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (bhagwat070919) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Thanks

**Comment by bhagwat070919** [comment]
Can you make changes and update the PR as per comments ?

**Review by bhks — CHANGES_REQUESTED** [review]

**Comment by Elias481** [comment]
> 
> 
> Can you make changes and update the PR as per comments ?

OK, only the one regarding the comment? Or some more?

**Comment by nckturner** [comment]
@Elias481 I'd like to see the desired behavior defined in tests as well.  It would be great to have that in this PR as well, just to make sure we all agree on "correct" behavior.

**Comment by bhagwat070919** [comment]
BTW , @Elias481  This fixes a really great issue which has been abandoning people from using the 
annotation `service.beta.kubernetes.io/aws-load-balancer-security-groups`.

I would be happy to ship this PR as soon as possible.

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (Elias481) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Duplicate code and inconsitentecy.
Of course if the duplication is desired here, then we should also check whether securityGroupID is from annotations to make it consitent.
Or to have less duplication, make another function that is called from both functions and restructure the buildELBSecurityGroupList according to that? Maybe even take out the setupSG part again and cann another function that is also called from setSecurityGroupIngress?
At the end setSecurityGroupIngress was not called anymore and the removed check (implemented wit setupSG flag) is done around the block that calls setSecurityGroupIngress (line 3859) so I would not double check but leave the setSecurityGroupIngress geenric - as I did.

**Comment by Elias481** [comment]
> 
> 
> BTW , @Elias481 This fixes a really great issue which has been abandoning people from using the
> annotation `service.beta.kubernetes.io/aws-load-balancer-security-groups`.
> 
> I would be happy to ship this PR as soon as possible.

I did not found time yesterday. So I can collect the requested changes from You @bhagwat070919  (only comment ore more?) and check the test requested again by by @nckturner (the topic was also in discussion with @wongma7 where he told he could add the durther tests later) - but yesterday without time I did not understood my own new test which already should cover one setupSG-positive and one -negative case. This is a least requirement for me so I would like to understand again why and whether it works.

**Comment by nckturner** [comment]
I'd like a couple things from you @Elias481, can you update **"What this PR does / why we need it"** to be detailed (someone looking at this PR should not have to follow links to understand the intent of the change).  

For example, as the PR is written today, `aws-load-balancer-security-groups` will replace the default security groups and not be modified, whereas any extra security groups provided in `aws-load-balancer-extra-security-groups` is always appended and will sometimes be modified.  If I understand correctly, then I think this behavior is still wrong so I want to understand your reasoning here. (are you trying to preserve backwards compatibility in this case?)

Additionally, once we decide on the intended behavior for both annotations, and whether or not we are introducing a breaking change for any users, then we should make sure to update docs and add tests.  I agree with @wongma7 though, that can be in a future PR.

**Comment by nckturner** [comment]
> is always appended and will sometimes be modified.

Just had another look at the code and saw that we pass the first SG in the list, which currently is either the default SG or could be from `aws-load-balancer-security-groups`: 

 `_, err = c.setSecurityGroupIngress(securityGroupIDs[0], permissions)`

So I don't think your proposed behavior is wrong, but would still appreciate it if you filled out the issue summary with more detail.

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (nckturner) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: Might be more clear to return the default SG separately instead of the bool, that way you can check for its existence and its obvious that we're only passing that one to `setSecurityGroupIngress().`

**Review by nckturner — CHANGES_REQUESTED** [review]

**Comment by bhagwat070919** [comment]
I think we should consider the 1.18 release date and try to wrap up this one to get it merged in the release if we want to.

For changes @Elias481  yes just the comment and the global Security group check removed code , I am not sure about the behavior after not having that check  we definitely need better testing in place but for now can we keep it there so that we are not introducing any other Bug.

I like the idea of returning the securityGroupID instead of boolean as to be specific , incase someone just reordered the List update so index 0 will no longer be the positional value of default SG.

So the idea of @nckturner would help here as to return the "securityGroupID" and check for existance and enter the if statement and also only do below:

```
_, err = c.setSecurityGroupIngress(securityGroupID, permissions)

``` 
instead of 

```
_, err = c.setSecurityGroupIngress(securityGroupIDs[0], permissions)

```

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (bhagwat070919) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: @nckturner ,

But here is the problem with that in case of SG is coming from c.cfg.Global.ElbSecurityGroup we would not be able to understand the difference in the calling function and based on the behavior we don't want to change the Ingress of Global SG.

**Comment by nckturner** [comment]
@Elias481 I spoke with @bhagwat070919 and we are ok with the PR as is, we would prefer it to be merged before next week.  Further testing and refactoring the confusing way we pass around the list of security groups can come in follow up PR.  

/lgtm

**Comment by Elias481** [comment]
OK fine, I updated the description a bit. So hopefully we can get the fixed version soon in our live enviroment.

**Comment by bhagwat070919** [comment]
@Elias481 

We should update the release notes so that it can be added in next release.

**Comment by Elias481** [comment]
@bhagwat070919 shall I do another PR for that or how? I think it's just a short note that the behaviour that was probably expected/desired is now in place.

**Comment by ivan-sukhomlyn** [comment]
@Elias481 @bhagwat070919 Hi guys,
After this PR merging, will Kubernetes continue to create its own Security Group in case if `aws-load-balancer-security-groups` defined in the annotations list with custom SG?
More details are available by the link https://github.com/kubernetes/cloud-provider-aws/issues/65
Thanks

**Comment by Elias481** [comment]
@ivan-sukhomlyn You are right, I forgot to mention and possibly worth to adding to description of PR und Release notes explicitly. This is of course also fixed in this turn.

**Comment by ivan-sukhomlyn** [comment]
Thanks for clarifying and your work 👍 
@bhagwat070919 Maybe you have some ETA of Kubernetes 1.18 at the AWS EKS service?

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go
Comment (seh) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: s/setupSg/setUpSG/

**Comment by bhagwat070919** [comment]
> @bhagwat070919 shall I do another PR for that or how? I think it's just a short note that the behaviour that was probably expected/desired is now in place.

@Elias481  you can just update this PR Release note and it would be taken when release note is being made.

**Comment by bhagwat070919** [comment]
@Elias481 ,

Do you mind creating Cherry-pick of this PR for k8s 1.17, 1.16, 1.15 and 1.14 ?

**Comment by Elias481** [comment]
@bhagwat070919 I can do so, I think I will find some time till monday morning..

**Comment by ivan-sukhomlyn** [comment]
@Elias481 @bhagwat070919 It would be nice to have this fix in other K8S versions.

**Comment by vampirismtrueblood** [comment]
I think this issue needs to be re-opened, issue still exists with EKS v1.25, 
loadBalancerSourceRanges works when creating the NLB, but updating still does not work.
Can we have this re-opened please

**Pull request merged**
